pkg/test/echo/server/endpoint/tcp.go

```diff
@@ -75,23 +75,27 @@ func (s *tcpInstance) Start(onReady OnReadyFunc) error {
 }
 
 // Handles incoming connection.
-func (s *tcpInstance) echo(conn io.ReadWriteCloser) {
+func (s *tcpInstance) echo(conn net.Conn) {
 	defer common.Metrics.TCPRequests.With(common.Port.Value(strconv.Itoa(s.Port.Port))).Increment()
 	defer func() {
 		_ = conn.Close()
 	}()
 
-	// Fill the field in the response
-	_, _ = conn.Write([]byte(fmt.Sprintf("%s=%s\n", string(response.StatusCodeField), response.StatusCodeOK)))
-
+	initialReply := true
 	for {
 		buf, err := bufio.NewReader(conn).ReadBytes(byte('\n'))
 		if err != nil {
 			if err != io.EOF {
-				log.Warn("tcp read failed: " + err.Error())
+				log.Warn("TCP read failed: " + err.Error())
 			}
 			return
 		}
+		log.Infof("TCP Request:\n  Source IP:%s\n  Destination Port:%d", conn.RemoteAddr(), s.Port.Port)
+		if initialReply {
```

Comment:
i may lack the background. but why we need to handle the first read of tcp data differently, i.e. relying some status status=ok? is there some test code built on top of this behavior? seems to be the envoy issue only require us not to reply right after the tcp connection is established, right?

Reply:
and what if we don't have this initial Reply check, but do status message write always in the loop body.

Reply:
> why we need to handle the first read of tcp data differently, i.e. relying some status status=ok? is there some test code built on top of this behavior?

I don't think anything is relying on this.

> seems to be the envoy issue only require us not to reply right after the tcp connection is established, right?

Yes, we just need to avoid writing data on the very first connection establishment, putting it behind read() guarantee this otherwise it will somehow depends on the timing or caching, making the test flaky.

Reply:
> what if we don't have this initial Reply check, but do status message write always in the loop body.

That logic is added by @yxue so I guess it will break something otherwise.

Reply:
ok. i just want to understand the issue and the fix.

Reply:
The initial Reply was added to make the client check happy and clean. The echo forwarder extracts several fields from the response and returns to the client. Adding the status line will make the http and tcp check consistent.